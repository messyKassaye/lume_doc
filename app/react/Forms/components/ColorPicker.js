import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { TwitterPicker } from 'react-color';

import { COLORS } from 'app/utils/colors';

class ColorPicker extends Component {
  constructor(props) {
    super(props);
    this.state = { active: false };
    this.onButtonClick = this.onButtonClick.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onColorChange = this.onColorChange.bind(this);
  }

  onColorChange({ hex }) {
    const { onChange } = this.props;
    onChange(hex);
  }

  onButtonClick() {
    this.setState(oldState => ({ active: !oldState.active }));
  }

  onBlur() {
    this.setState({ active: false });
  }

  render() {
    const { active } = this.state;
    const { value, defaultValue } = this.props;
    return (
      <div className="ColorPicker">
        <div
          className="ColorPicker__button"
          style={{ backgroundColor: value || defaultValue }}
          onClick={this.onButtonClick}
        />
        {active && (
          <div className="ColorPicker__popover">
            <div className="ColorPicker__cover" onClick={this.onBlur} />
            <TwitterPicker
              color={value || defaultValue}
              colors={COLORS}
              onChangeComplete={this.onColorChange}
            />
          </div>
        )}
      </div>
    );
  }
}

ColorPicker.defaultProps = {
  value: '',
  defaultValue: '',
};

ColorPicker.propTypes = {
  value: PropTypes.string,
  defaultValue: PropTypes.string,
  onChange: PropTypes.func.isRequired,
};

export default ColorPicker;
