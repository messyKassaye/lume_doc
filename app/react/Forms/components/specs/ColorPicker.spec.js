import React from 'react';
import { shallow } from 'enzyme';
import { TwitterPicker } from 'react-color';

import ColorPicker from '../ColorPicker';

describe('ColorPicker', () => {
  let component;
  let props;

  beforeEach(() => {
    props = {
      defaultValue: '#112233',
      onChange: jest.fn(),
    };
  });

  const render = () => {
    component = shallow(<ColorPicker {...props} />);
  };

  const renderAndActivate = () => {
    render();
    component.setState({ active: true });
    component.update();
  };

  it('should render ColorPicker button with specified value as color', () => {
    props.value = '#ffffff';
    render();
    expect(component).toMatchSnapshot();
  });

  it('should show defaultValue color if value is not provided', () => {
    render();
    expect(component).toMatchSnapshot();
  });

  describe('when button is clicked', () => {
    it('should display color picker if hidden', () => {
      render();
      component
        .find('.ColorPicker__button')
        .first()
        .simulate('click');
      component.update();
      expect(component).toMatchSnapshot();
    });
    it('should hide color picker if displayed', () => {
      renderAndActivate();
      component
        .find('.ColorPicker__button')
        .first()
        .simulate('click');
      expect(component).toMatchSnapshot();
    });
  });

  describe('when color picker is displayed', () => {
    it('should close when clicking outside', () => {
      renderAndActivate();
      component
        .find('.ColorPicker__cover')
        .first()
        .simulate('click');
      expect(component).toMatchSnapshot();
    });
  });

  it('should call onChange with when new color is selected', () => {
    renderAndActivate();
    component.find(TwitterPicker).prop('onChangeComplete')({ hex: '#001100' });
    expect(props.onChange).toHaveBeenCalledWith('#001100');
  });
});
