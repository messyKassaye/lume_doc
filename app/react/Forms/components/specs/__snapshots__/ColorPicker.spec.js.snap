// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker should render ColorPicker button with specified value as color 1`] = `
<div
  className="ColorPicker"
>
  <div
    className="ColorPicker__button"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  />
</div>
`;

exports[`ColorPicker should show defaultValue color if value is not provided 1`] = `
<div
  className="ColorPicker"
>
  <div
    className="ColorPicker__button"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#112233",
      }
    }
  />
</div>
`;

exports[`ColorPicker when button is clicked should display color picker if hidden 1`] = `
<div
  className="ColorPicker"
>
  <div
    className="ColorPicker__button"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#112233",
      }
    }
  />
  <div
    className="ColorPicker__popover"
  >
    <div
      className="ColorPicker__cover"
      onClick={[Function]}
    />
    <ColorPicker
      color="#112233"
      colors={
        Array [
          "#C03B22",
          "#D9534F",
          "#E91E63",
          "#A03AB1",
          "#6F46B8",
          "#3F51B5",
          "#2196F3",
          "#37BDCf",
          "#359990",
          "#5CB85C",
          "#8BC34A",
          "#CDDC39",
          "#CCBC2F",
          "#F0AD4E",
          "#EC9920",
          "#E46841",
          "#795548",
          "#9E9E9E",
          "#607D8B",
        ]
      }
      onChangeComplete={[Function]}
      styles={Object {}}
      triangle="top-left"
      width={276}
    />
  </div>
</div>
`;

exports[`ColorPicker when button is clicked should hide color picker if displayed 1`] = `
<div
  className="ColorPicker"
>
  <div
    className="ColorPicker__button"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#112233",
      }
    }
  />
</div>
`;

exports[`ColorPicker when color picker is displayed should close when clicking outside 1`] = `
<div
  className="ColorPicker"
>
  <div
    className="ColorPicker__button"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#112233",
      }
    }
  />
</div>
`;
